core/unit_test/tools/TestEventCorrectness.hpp

```diff
@@ -159,6 +159,7 @@ TEST(defaultdevicetype, test_unnamed_instance_fence) {
         {"Unnamed Instance Fence", FencePayload::distinguishable_devices::no,
          root + num_instances}};
     expect_fence_events(expected, [=]() {
+      // TODO: This is the global instance for HPX. Should it not be?
```

Comment:
A few more thoughts on this. We now have code for partitioning an instance. In CUDA-land, this means creating streams. Consider the following code

```c++
    Kokkos::DefaultExecutionSpace default_space;
    auto instances = Kokkos::Experimental::partition_space(default_space, 1, 1);
    view_type temperature_field1("temperature_one", data_size);
    view_type temperature_field2("temperature_two", data_size);
    auto f1_mirror = Kokkos::create_mirror_view(temperature_field1);
    auto f2_mirror = Kokkos::create_mirror_view(temperature_field2);
    for (int x = 0; x < repeats; ++x) {
      Kokkos::parallel_for(
          "process_temp1", Kokkos::RangePolicy<Kokkos::DefaultExecutionSpace>(instances[0], 0, data_size),
          KOKKOS_LAMBDA(int i) { temperature_field1(i) -= 1.0f; });
      Kokkos::deep_copy(instances[0], f1_mirror, temperature_field1);
      Kokkos::parallel_for(
          "process_temp2", Kokkos::RangePolicy<Kokkos::DefaultExecutionSpace>(instances[1], 0, data_size),
          KOKKOS_LAMBDA(int i) { temperature_field2(i) -= 1.0f; });
      Kokkos::deep_copy(instances[1],f2_mirror, temperature_field2);
```

Now, in a CUDA case, if I were to add a fence on "root_space" somewhere in there, it wouldn't fence the instances that were split off (and this is what we considered desirable). In HPX, with your current semantics, fencing the default space fences everything. If I were writing a Kokkos backend, I'd try to make my semantics mirror those of the most-used backends, which in the asynchronous case would be the GPU backends.

It's not a "rule" of Kokkos to my knowledge (@crtrott could say more on that), but if I was playing your hand, I would not have special semantics for the default instance

Reply:
Thanks for the explanation, that makes sense. I haven't really looked at the partitioning so far. What we're using is a flag to the HPX constructor to create an independent instance. It looks like it would be quite easy to implement `partition_space` based on that for HPX. You also don't have to put in much effort to convince me that fencing the default instance should only fence itself. It makes my life easier as well. I think the main reason it fences everything right now is because the default Cuda instance used to do `cudaDeviceSynchronize` (or do I completely misremember...?). Anyways, what it used to do doesn't matter too much. I'll update it to fence itself only.

Reply:
I think the instance ids and fences now behave reasonably. I've left the event correctness test unchanged and it passes so I take that as a sign that it now behaves as you (and I) would hope and expect.